### PR TITLE
Rewrite grader-host with async/await

### PR DESCRIPTION
**Pull request by nwalters512**
This started out as an effort to get rid of the last usages of `async-stacktrace`, which of course quickly morphed into a large refactoring of `grader-host` 😅 this is probably the 4th time I've attempted this, and I think I got it right this time. I've tested this in staging and it works correctly.

### CODE REVIEW

File: apps/grader-host/src/index.js [hunk 1]
```diff
@@ -457,311 +384,238 @@ function runJob(info, callback) {
...
-          (err) => {
-            if (ERR(err, callback)) return;
-            callback(null);
-          },
-        );
-      },
-      (callback) => {
-        // We made it through the Docker danger zone!
-        clearTimeout(jobTimeoutId ?? undefined);
-        jobTimeoutId = null;
```
Comment (nwalters512) [inline apps/grader-host/src/index.js:575]: Calling this out for review: AFAICT, we don't actually need to null this out for the purposes of avoiding calling `clearTimeout` multiple times. Node will silently do nothing on subsequent calls on `clearTimeout` with the same timer object.

File: apps/grader-host/src/index.js [hunk 2]
```diff
@@ -457,311 +384,238 @@ function runJob(info, callback) {
...
       // the timeout here if needed.
-      if (jobTimeoutId != null) {
-        clearTimeout(jobTimeoutId);
-      }
-
-      // If we somehow eventually get here after exceeding the global timeout,
-      // we should avoid calling the callback again
-      if (jobFailed) {
-        return;
-      }
```
Comment (nwalters512) [inline apps/grader-host/src/index.js:632]: This is no longer necessary. We now use `Promise.race` to concurrently execute the timeout and the grading job, so if the timeout finishes first, it doesn't matter that the grading job promise resolves later; the result will just be ignored.

File: apps/grader-host/src/lib/deferred.js
Comment (nwalters512) [inline apps/grader-host/src/lib/deferred.js:1]: Can't wait for our Node 22 upgrade: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers

**Review by mylesw27 — APPROVED** [review]
Walked through the code with @nwalters512 on Zoom. I am unable to test locally but Nathan has tested and everything works.

**Comment by nwalters512** [comment]
I tested this one more time in our staging environment and everything is looking good 🎉

**Pull request merged**